mc-sema/runtime/linux_i386_callback.c

```diff
@@ -0,0 +1,728 @@
+
+    // save return value into rax
+    __asm__ volatile("movl %0, %%eax\n": : "m"(__mcsema_do_call_state.EAX) );
+
+    __asm__ volatile("retl\n");
+}
+
+typedef struct _do_call_state_t {
+    uint32_t __mcsema_real_esp;
+    uint32_t __mcsema_jmp_count;
+    char sse_state[512] __attribute__((aligned (16)));
+    uint32_t reg_state[15];
+} do_call_state_t;
+
+__thread do_call_state_t do_call_state[NUM_DO_CALL_FRAMES];
+__thread int32_t cur_do_call_frame = -1; /* XXX */
+__thread uint32_t call_frame_counter = 0; /* XXX */
+
+void do_call_value(void *state, uint32_t value)
+{
+    // get a clean frame to store state
+    ++cur_do_call_frame;
+    do_call_state_t *cs = &(do_call_state[cur_do_call_frame]);
+    call_frame_counter = 0;
+    cs->__mcsema_jmp_count = NUM_DO_CALL_FRAMES - cur_do_call_frame - 1;
+    //uint32_t reg_state[] = cs->reg_state;
+
+    __asm__ volatile(
+            "pusha\n" // save all regs just so we don't have bother keeping track of what we saved
+            "fxsave %0\n" // save sse state
+            "movl %3, %%eax\n"  // capture "state" arg (mcsema regstate)
+            "movl %4, %%ecx\n"  // capture "value" arg (call destination)
+            "movl %2, %%esi\n" // pointer to TLS area where we save state
+            "movl %c[state_edi](%%eax), %%edi\n" // dump struct regs to state
+            "movl %c[state_edx](%%eax), %%edx\n"
+            "movl %c[state_ebx](%%eax), %%ebx\n"
+            "movl %c[state_ebp](%%eax), %%ebp\n"
+            "movups %c[state_xmm0](%%eax), %%xmm0\n" // dump struct regs xmm state
+            "movups %c[state_xmm1](%%eax), %%xmm1\n"
+            "movups %c[state_xmm2](%%eax), %%xmm2\n"
+            "movups %c[state_xmm3](%%eax), %%xmm3\n"
+            "movups %c[state_xmm4](%%eax), %%xmm4\n"
+            "movups %c[state_xmm5](%%eax), %%xmm5\n"
+            "movups %c[state_xmm6](%%eax), %%xmm6\n"
+            "movups %c[state_xmm7](%%eax), %%xmm7\n"
+            "leal %c[real_esp_off](%%esi), %%esi\n" // where will we save the "real" esp?
+            "movl %%esp, (%%esi)\n" // save our esp since we will switch to mcsema esp later
+            "movl %c[state_esp](%%eax), %%esp\n" // switch to mcsema stack
+            "addl $4, %%esp\n" // we pushed a fake return addr before, undo it
+            "movl %%ecx, -4(%%esp)\n" // use that slot to store jump destination
+            "movl %c[jmp_count](%%esi), %%ecx\n" // save recursion count into ecx
+            "imull $7, %%ecx, %%ecx\n" //use that to calc how many INC instructions to skip
+            "movl 0f, %%esi\n" // base return addr
+            "addl %%ecx, %%esi\n" // calculate return addr
+            "pushl %%esi\n" // push return addr
+            "movl %c[state_ecx](%%eax), %%ecx\n" // complete struct regs spill
+            "movl %c[state_esi](%%eax), %%esi\n"
+            "movl %c[state_eax](%%eax), %%eax\n"
+            "jmpl *-4(%%esp)\n"
```

Comment:
we added 4 to ESP previously since there was a fake return address there, so the memory slot should be valid. I was debating doing push/ret, but figured a jump was more obvious.